File: apps/api/app/schemas/billing.py
"""Pydantic schemas for the billing surface."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.billing import PlanTier, SubscriptionStatus, UsageKind


class CheckoutRequest(BaseModel):
    plan: PlanTier
    seats: int = Field(default=1, ge=1, le=10_000)
    success_url: str
    cancel_url: str


class CheckoutResponse(BaseModel):
    url: str


class PortalRequest(BaseModel):
    return_url: str


class PortalResponse(BaseModel):
    url: str


class SubscriptionRead(BaseModel):
    plan: PlanTier
    status: SubscriptionStatus
    seats: int
    trial_ends_at: datetime | None = None
    current_period_start: datetime | None = None
    current_period_end: datetime | None = None
    cancel_at_period_end: bool = False

    model_config = ConfigDict(from_attributes=True)


class UsageSummaryEntry(BaseModel):
    kind: UsageKind
    used: int
    limit: int  # 0 = unlimited
    remaining: int  # -1 = unlimited


class UsageSummary(BaseModel):
    plan: PlanTier
    period_start: datetime
    entries: list[UsageSummaryEntry]

File: apps/web/src/lib/api/billing.ts
import { api } from "./client";

export type PlanTier = "free" | "pro" | "team" | "business" | "enterprise";

export type SubscriptionStatus =
  | "trialing"
  | "active"
  | "past_due"
  | "canceled"
  | "incomplete"
  | "paused";

export type UsageKind =
  | "transcription_minutes"
  | "llm_input_tokens"
  | "llm_output_tokens"
  | "bot_minutes"
  | "storage_gb_hours"
  | "seats";

export interface PlanCatalogEntry {
  name: string;
  monthly_price_cents: number;
  seat_limit: number;
  features: string[];
  limits: Partial<Record<UsageKind, number>>;
}

export interface Subscription {
  plan: PlanTier;
  status: SubscriptionStatus;
  seats: number;
  trial_ends_at: string | null;
  current_period_start: string | null;
  current_period_end: string | null;
  cancel_at_period_end: boolean;
}

export interface UsageEntry {
  kind: UsageKind;
  used: number;
  limit: number; // 0 = unlimited
  remaining: number; // -1 = unlimited
}

export interface UsageReport {
  plan: PlanTier;
  period_start: string;
  entries: UsageEntry[];
}

export const billingApi = {
  plans: () => api.get<Record<PlanTier, PlanCatalogEntry>>("/api/v1/billing/plans"),
  subscription: () => api.get<Subscription>("/api/v1/billing/subscription"),
  usage: () => api.get<UsageReport>("/api/v1/billing/usage"),

  checkout: (body: {
    plan: PlanTier;
    seats?: number;
    success_url: string;
    cancel_url: string;
  }) => api.post<{ url: string }>("/api/v1/billing/checkout", body),

  portal: (return_url: string) =>
    api.post<{ url: string }>("/api/v1/billing/portal", { return_url }),
};
